src/lib/core/ripple/ripple-renderer.ts

```diff
@@ -171,21 +171,15 @@ export class RippleRenderer {
       this._mostRecentTransientRipple = rippleRef;
     }
 
-    // Wait for the ripple element to be completely faded in.
-    // Once it's faded in, the ripple can be hidden immediately if the mouse is released.
-    this.runTimeoutOutsideZone(() => {
-      const isMostRecentTransientRipple = rippleRef === this._mostRecentTransientRipple;
-
-      rippleRef.state = RippleState.VISIBLE;
+    this._ngZone.runOutsideAngular(() => {
+      ripple.addEventListener('transitionend', () => this._onRippleTransitionEnd(rippleRef));
```

Comment:
We've had issues in the past with `transitionend` not firing reliably; in [OverlayRef](https://github.com/angular/material2/blob/71886f84efa0fb3c1e8ac66e13807dfd17fa2848/src/cdk/overlay/overlay-ref.ts#L404-L415) we use both `transitionend` and a fallback `setTimeout` for cases when the event doesn't fire.

@crisbeto might remember more context about that case

Reply:
`transitionend` won't fire if the element doesn't have a transition, or if a transition is too short. E.g. I had something similar on the drag&drop where the even wouldn't fire if you hadn't dragged the element far enough.

Reply:
Yeah that's why I have an additional check that checks if the `duration` is zero. In that case we still want to keep the event registered because there can be a different duration for the `exit` animation.

We could technically also skip registering `transitionEnd` if both enter and exit duration is equal to zero. So we save some memory and avoid an unnecessary listener.